### ISSUE TITLE
Outgoing traffic is slow when sent outside a-shell

### ISSUE THREAD

**Issue by Crissium** [holzschu/a-shell#707]
I am developing a Python [program](https://github.com/Crissium/SilverDict) that starts a server listening at localhost:2628. It runs without any problems in a-shell, and when queried with curl locally, responds successfully and the speed is very good. However, if accessed from the local Safari or with curl on another device, the speed is intolerably slow.

```bash
time curl 'http://10.22.135.214:2628/api/query/Default%20Group/word'
^C

real	6m34.410s
user	0m0.024s
sys	0m0.000s
```

(Safari actually loads the page at last, but no luck with curl.)

The strange thing is that only scanning the port with netcat is also instantaneous:

```bash
time nc -zv 10.22.135.214 2628
Connection to 10.22.135.214 2628 port [tcp/dict] succeeded!

real	0m0.189s
user	0m0.002s
sys	0m0.000s
```

So what is the reason of this problem? And if the speed could not be improved, could you please add support for w3m or other command-line browsers?

**Comment by holzschu** [comment]
I'll have to run tests to confirm, but my first hypothesis is that iOS does not like running apps in the background. In order to run a web server inside a-Shell and access it from Safari, you need both apps to be in the foreground (which is possible on an iPad). The same applies when accessing from another device: you need both devices to be open and activated.
There could be other issues, as well.

**Comment by Crissium** [comment]
Thank you for your responsiveness! I have tested again with the terminal in the foreground, and this time it is extremely fast:

```
real	0m0.111s
user	0m0.000s
sys	0m0.009s
```

So could you please tell me how to keep both apps in the foreground? I am testing on an iPad and do not know my way round iOS.

**Comment by holzschu** [comment]
I'm glad the issue is resolved. I'll still have a look at w3m, it would be nice to have a text-based browser inside a-Shell.

**Comment by holzschu** [comment]
Hi, the latest TestFlight version (https://testflight.apple.com/join/WUdKe3f4 for a-Shell, https://testflight.apple.com/join/REdHww5C for a-Shell mini) has a brand new internal web browser (internalbrowser) that opens the URL in an a-Shell window, which lets your Python server continue running but shows the webview.
You can call it either by its name (internalbrowser url) or by setting the environment variable BROWSER to internalbrowser. Python will call the command BROWSER if it is set. The command openurl will open the browser defined by BROWSER.

You can navigate back and forward with left-edge-swipe and right-edge-swipe, so you can go back to the terminal.

It's not perfect yet: I'd like to have buttons at the top (but I don't see how to do it), you don't always get keyboard control when you go back to the terminal, and there is no output while the terminal is masked by the web view. For the last point, if you have two windows side by side on an iPad (with newWindow), the webview will likely be on the second window, so you can have both web view and output.

**Comment by Crissium** [comment]
Hi, holzschu. I cannot thank you enough for going the extra mile and implemented this feature so quickly! This internal web browser is very important for all server programs running in a-Shell, for example, mine, and this one for classicists: https://github.com/pjheslin/diogenes

However, the iPad on which I first tested my dictionary server was borrowed. But I shall immediately test this new version as soon as I have access to it again. Thanks again!

**Comment by Crissium** [comment]
> You can call it either by its name (internalbrowser url) or by setting the environment variable BROWSER to internalbrowser. Python will call the command BROWSER if it is set. The command openurl will open the browser defined by BROWSER.

Hello, I can confirm that `internalbrowser https://example.com` does open a webview, but if I first `export BROWSER=internalbrowser` and then run `python server.py`, only the output of the Python server shows up. Have I got it wrong? I am now testing on an iPhone 12.

And how should I keep a command running in the background? `python server.py &` does not seem to work…

Thanks again for your effort!

**Comment by holzschu** [comment]
I think it depends on what command your Python script uses to launch a browser. If it uses the `webbrowser` module, it will call the content of the `BROWSER` environment variable.

You cannot run a command in the background, but if your Python script calls `internalbrowser`, then the server and the client will be running in parallel.

**Comment by Crissium** [comment]
Thanks for explaining! I have never heard of the module and will modify my code and try again.

**Comment by Crissium** [comment]
I am happy to report back that the internal browser works very well with my Python server. Thank you very much for implementing this!

**Comment by holzschu** [comment]
By the way, #736 is also about running Diogenes and using the internalbrowser (but the Perl version).
